x/ibc/04-channel/keeper/packet.go

```
@@ -77,8 +77,24 @@ func (k Keeper) SendPacket(
 	}
 
 	// check if packet timeouted on the receiving chain
-	if clientState.GetLatestHeight() >= packet.GetTimeoutHeight() {
-		return sdkerrors.Wrap(types.ErrPacketTimeout, "timeout already passed ond the receiving chain")
+	latestHeight := clientState.GetLatestHeight()
+	if packet.GetTimeoutHeight() != 0 && latestHeight >= packet.GetTimeoutHeight() {
+		return sdkerrors.Wrapf(
+			types.ErrPacketTimeout,
+			"receiving chain block height >= packet timeout height (%d >= %d)", latestHeight, packet.GetTimeoutHeight(),
+		)
+	}
+
+	latestTimestamp, err := k.connectionKeeper.GetTimestampAtHeight(ctx, connectionEnd, latestHeight)
+	if err != nil {
+		return err
+	}
+
+	if packet.GetTimeoutTimestamp() != 0 && latestTimestamp >= packet.GetTimeoutTimestamp() {
+		return sdkerrors.Wrapf(
+			types.ErrPacketTimeout,
+			"receiving chain block timestamp >= packet timeout timestamp (%d >= %d)", latestTimestamp, packet.GetTimeoutTimestamp(),
```

nit, I'd return the time errors using `time.Time` instead of `uint64`